src/transformStateWithClones.js

```diff
@@ -6,8 +6,27 @@
  *
  * @return {Object[]}
  */
+
 function transformStateWithClones(state, actions) {
-  // write code here
+  const stateHistory = [];
+  let currentState = { ...state };
+
+  for (const action of actions) {
+    if (action.type === 'clear') {
+      currentState = {};
+    } else if (action.type === 'addProperties') {
+      currentState = { ...currentState, ...action.extraData };
+    } else if (action.type === 'removeProperties') {
+      currentState = { ...currentState };
+
+      for (const key of action.keysToRemove) {
+        delete currentState[key];
+      }
+    }
+    stateHistory.push(currentState);
```

Comment:
To avoid repetition and enhance clarity, push the `currentState` to `stateHistory` after the switch statement, outside of it, once the current action has been processed. This adheres to the DRY (Don't Repeat Yourself) principle.